sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala

```diff
@@ -334,15 +336,19 @@ class JacksonParser(
   private def convertObject(
       parser: JsonParser,
       schema: StructType,
-      fieldConverters: Array[ValueConverter]): InternalRow = {
+      fieldConverters: Array[ValueConverter],
+      structFilters: StructFilters = new NoopFilters()): Option[InternalRow] = {
     val row = new GenericInternalRow(schema.length)
     var badRecordException: Option[Throwable] = None
+    var skipRow = false
 
-    while (nextUntil(parser, JsonToken.END_OBJECT)) {
+    structFilters.reset()
+    while (!skipRow && nextUntil(parser, JsonToken.END_OBJECT)) {
       schema.getFieldIndex(parser.getCurrentName) match {
         case Some(index) =>
           try {
             row.update(index, fieldConverters(index).apply(parser))
+            skipRow = structFilters.skipRow(row, index)
```

Comment:
Hm, if I didn't misread, this wouldn't work properly (with https://github.com/apache/spark/pull/27366/files#diff-635e02b2d1ce4ad1675b0350ccac0c10R266) if we happen to push down predicates that refer nested columns. 

Seems a bit odd because the previous code didn't have such assumption before and all codes were valid for both root and nested rows. Now this method works differently for root row and nested row. I wonder if we can decouple the filtering logic out of this method ...

Reply:
> if we happen to push down predicates that refer nested columns.

I did some experiments, and found out that filters with nested attributes are not pushed down.

Reply:
@dbtsai is trying to support them https://github.com/dbtsai/spark/pull/10

Reply:
> Now this method works differently for root row and nested row. I wonder if we can decouple the filtering logic out of this method ...

Initially, I just duplicate the code but in the commit collapsed it to one method: https://github.com/apache/spark/pull/27366/commits/94a22e1cbbc3fc5ca5d87f6b08c8b5adc121a7b8